From 82eeb7da67b14add30645d2d2c1740f9a4b43723 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Mon, 10 Aug 2020 07:30:27 -0700
Subject: [PATCH] add a interface to skip automount devtmpfs to /dev/ path

Upstream-Status: Inappropriate [other]

systemd will automount devtmpfs to /dev/, but when run VRF, it will
cause problem;

VRF used lxc to create a container; to log on the container by
(lxc-connect), the /dev/tty1 and /dev/pts/1 must be created, and be
bound when start a container by lxc-start.

but the bound tty device will be overwritten when systemd start since
system will remount /dev/.

Add a interface, which is to check if /.skip_mount_devtmpfs_on_dev
exist, to skip mount devtmpfs to /dev/ path

Signed-off-by: Roy Li <rongqing.li@windriver.com>

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Rebase to systemd 247
Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 src/core/mount-setup.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/core/mount-setup.c b/src/core/mount-setup.c
index 915b1016b8..2452b1ab21 100644
--- a/src/core/mount-setup.c
+++ b/src/core/mount-setup.c
@@ -61,13 +61,15 @@ typedef struct MountPoint {
 #define N_EARLY_MOUNT 4
 #endif
 
+bool skip_mount_dev();
+
 static const MountPoint mount_table[] = {
         { "proc",        "/proc",                     "proc",       NULL,                                      MS_NOSUID|MS_NOEXEC|MS_NODEV,
           NULL,          MNT_FATAL|MNT_IN_CONTAINER|MNT_FOLLOW_SYMLINK },
         { "sysfs",       "/sys",                      "sysfs",      NULL,                                      MS_NOSUID|MS_NOEXEC|MS_NODEV,
           NULL,          MNT_FATAL|MNT_IN_CONTAINER },
         { "devtmpfs",    "/dev",                      "devtmpfs",   "mode=755" TMPFS_LIMITS_DEV,               MS_NOSUID|MS_NOEXEC|MS_STRICTATIME,
-          NULL,          MNT_FATAL|MNT_IN_CONTAINER },
+          skip_mount_dev,          MNT_FATAL|MNT_IN_CONTAINER },
         { "securityfs",  "/sys/kernel/security",      "securityfs", NULL,                                      MS_NOSUID|MS_NOEXEC|MS_NODEV,
           NULL,          MNT_NONE                   },
 #if ENABLE_SMACK
@@ -125,6 +127,12 @@ bool mount_point_is_api(const char *path) {
         return path_startswith(path, "/sys/fs/cgroup/");
 }
 
+bool skip_mount_dev() {
+       struct stat buf;
+
+       return stat("/.skip_mount_devtmpfs_on_dev", &buf) == -1;
+}
+
 bool mount_point_ignore(const char *path) {
 
         const char *i;
-- 
2.17.1

